File: lang/java/ipc/src/main/java/org/apache/avro/ipc/Responder.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avro.ipc;

import org.apache.avro.AvroRuntimeException;
import org.apache.avro.Protocol;
import org.apache.avro.Protocol.Message;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericDatumReader;
import org.apache.avro.generic.GenericDatumWriter;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.Decoder;
import org.apache.avro.io.Encoder;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.specific.SpecificDatumReader;
import org.apache.avro.specific.SpecificDatumWriter;
import org.apache.avro.util.ByteBufferInputStream;
import org.apache.avro.util.ByteBufferOutputStream;
import org.apache.avro.util.Utf8;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutionException;
import java.util.function.BiConsumer;
import java.util.function.Function;

/** Base class for the server side of a protocol interaction. */
public abstract class Responder {
  private static final Logger LOG = LoggerFactory.getLogger(Responder.class);

  private static final Schema META =
      Schema.createMap(Schema.create(Schema.Type.BYTES));
  private static final GenericDatumReader<Map<String, ByteBuffer>>
      META_READER = new GenericDatumReader<Map<String, ByteBuffer>>(META);
  private static final GenericDatumWriter<Map<String, ByteBuffer>>
      META_WRITER = new GenericDatumWriter<Map<String, ByteBuffer>>(META);

  private static final ThreadLocal<Protocol> REMOTE =
      new ThreadLocal<Protocol>();

  private final Map<MD5, Protocol> protocols
      = new ConcurrentHashMap<MD5, Protocol>();

  private final Protocol local;
  private final MD5 localHash;
  protected final List<RPCPlugin> rpcMetaPlugins;

  protected Responder(Protocol local) {
    this.local = local;
    this.localHash = new MD5();
    localHash.bytes(local.getMD5());
    protocols.put(localHash, local);
    this.rpcMetaPlugins =
        new CopyOnWriteArrayList<RPCPlugin>();
  }

  /** Return the remote protocol.  Accesses a {@link ThreadLocal} that's set
   * around calls to {@link #respond(Protocol.Message, Object)}. */
  public static Protocol getRemote() { return REMOTE.get(); }

  /** Return the local protocol. */
  public Protocol getLocal() { return local; }

  /**
   * Adds a new plugin to manipulate per-call metadata.  Plugins
   * are executed in the order that they are added.
   * @param plugin a plugin that will manipulate RPC metadata
   */
  public void addRPCPlugin(RPCPlugin plugin) {
    rpcMetaPlugins.add(plugin);
  }

  /** Called by a server to deserialize a request, compute and serialize
   * a response or error. */
  public List<ByteBuffer> respond(List<ByteBuffer> buffers) throws IOException {
    return respond(buffers, null);
  }

  /** Called by a server to deserialize a request, compute and serialize a
   * response or error.  Transciever is used by connection-based servers to
   * track handshake status of connection. */
  public List<ByteBuffer> respond(List<ByteBuffer> buffers, Transceiver connection) throws IOException {
    try {
      return respondAsync(buffers, connection).get();
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      return handleException(e, new RPCContext(), null);
    } catch (ExecutionException e) {
      final Throwable cause = e.getCause();
      if (cause instanceof IOException)
        throw (IOException) cause;
      throw new IOException(cause);
    }
  }

  public CompletableFuture<List<ByteBuffer>> respondAsync(List<ByteBuffer> buffers,
                                                          Transceiver connection) {
    Decoder in = DecoderFactory.get().binaryDecoder(
        new ByteBufferInputStream(buffers), null);
    ByteBufferOutputStream bbo = new ByteBufferOutputStream();
    BinaryEncoder out = EncoderFactory.get().binaryEncoder(bbo, null);
    final RPCContext context = new RPCContext();
    final boolean wasConnected = connection != null && connection.isConnected();

    final Message m;
    final Object request;
    final List<ByteBuffer> handshake;
    try {
      Protocol remote = handshake(in, out, connection);
      out.flush();
      if (remote == null) {                     // handshake failed
        return CompletableFuture.completedFuture(bbo.getBufferList());
      }
      handshake = bbo.getBufferList();

      // read request using remote protocol specification
      context.setRequestCallMeta(META_READER.read(null, in));
      String messageName = in.readString(null).toString();
      if (messageName.equals(""))            // a handshake ping
        return CompletableFuture.completedFuture(handshake);

      Message rm = remote.getMessages().get(messageName);
      if (rm == null)
        return CompletableFuture.completedFuture(compileResult(handshake, handleException(new AvroRuntimeException("No such remote message: " + messageName), context, null), context));

      m = getLocal().getMessages().get(messageName);
      if (m == null)
        return CompletableFuture.completedFuture(compileResult(handshake, handleException(new AvroRuntimeException("No message named " + messageName + " in " + getLocal()), context, null), context));

      request = readRequest(rm.getRequest(), m.getRequest(), in);
      context.setMessage(rm);
      for (RPCPlugin plugin : rpcMetaPlugins) {
        plugin.serverReceiveRequest(context);
      }
      // create response using local protocol specification
      if ((m.isOneWay() != rm.isOneWay()) && wasConnected) {
        return CompletableFuture.completedFuture(compileResult(handshake, handleException(new AvroRuntimeException("Not both one-way: " + messageName), context, m), context));
      }

      REMOTE.set(remote);
    } catch (IOException e) {
      final CompletableFuture<List<ByteBuffer>> future = new CompletableFuture<List<ByteBuffer>>();
      future.completeExceptionally(e);
      return future;
    }

    return respondAsync(m, request)
        .thenApply(new Function<Object, List<ByteBuffer>>() {
          @Override
          public List<ByteBuffer> apply(Object response) {
            context.setResponse(response);
            if (m.isOneWay() && wasConnected)
              return null;
            ByteBufferOutputStream bbo = new ByteBufferOutputStream();
            BinaryEncoder out = EncoderFactory.get().binaryEncoder(bbo, null);
            try {
              out.writeBoolean(false);
              writeResponse(m.getResponse(), response, out);
              out.flush();
            } catch (IOException e) {
              Responder.<RuntimeException>throwException(e);
            }
            return bbo.getBufferList();
          }
        })
        .exceptionally(new Function<Throwable, List<ByteBuffer>>() {
          @Override
          public List<ByteBuffer> apply(Throwable throwable) {
            return futureExceptionally(throwable, context, m);
          }
        })
        .whenComplete(new BiConsumer<List<ByteBuffer>, Throwable>() {
          @Override
          public void accept(List<ByteBuffer> byteBufferList, Throwable throwable) {
            REMOTE.set(null);
            if (byteBufferList != null) {
              context.setResponsePayload(byteBufferList);
              for (RPCPlugin plugin : rpcMetaPlugins) {
                plugin.serverSendResponse(context);
              }
            }
          }
        })
        .thenApply(new Function<List<ByteBuffer>, List<ByteBuffer>>() {
          @Override
          public List<ByteBuffer> apply(List<ByteBuffer> byteBufferList) {
            try {
              return compileResult(handshake, byteBufferList, context);
            } catch (IOException e) {
              Responder.<RuntimeException>throwException(e);
              return null;
            }
          }
        });
  }

  private static List<ByteBuffer> compileResult(List<ByteBuffer> handshake, List<ByteBuffer> payload, RPCContext context) throws IOException {
    if (payload == null)
      return null;
    ByteBufferOutputStream bbo = new ByteBufferOutputStream();
    BinaryEncoder out = EncoderFactory.get().binaryEncoder(bbo, null);
    META_WRITER.write(context.responseCallMeta(), out);
    out.flush();
    bbo.prepend(handshake == null ? new ByteBufferOutputStream().getBufferList() : handshake);
    bbo.append(payload);
    return bbo.getBufferList();
  }


  private List<ByteBuffer> handleException(Exception exception, RPCContext context, Message m) throws IOException {
    ByteBufferOutputStream bbo = new ByteBufferOutputStream();
    BinaryEncoder out = EncoderFactory.get().binaryEncoder(bbo, null);
    out.writeBoolean(true);
    try {
      context.setError(exception);
      if (m != null)
        writeError(m.getErrors(), exception, out);
      else
        writeError(Protocol.SYSTEM_ERRORS, new Utf8(exception.toString()), out);
      out.flush();
      return bbo.getBufferList();
    } catch (Exception e) {
      LOG.warn("system error", e);
      context.setError(e);
      writeError(Protocol.SYSTEM_ERRORS, new Utf8(e.toString()), out);
      out.flush();
      return bbo.getBufferList();
    }
  }

  private List<ByteBuffer> futureExceptionally(Throwable throwable, RPCContext context, Message m) {
    try {
      LOG.warn("user error", throwable.getCause());
      return handleException((Exception) throwable.getCause(), context, m);
    } catch (IOException e) {
      Responder.<RuntimeException>throwException(e);
      return null;
    }
  }

  @SuppressWarnings("unchecked")
  private static <T extends Throwable> void throwException(Throwable exception) throws T {
    throw (T) exception;
  }

  private SpecificDatumWriter<HandshakeResponse> handshakeWriter =
      new SpecificDatumWriter<HandshakeResponse>(HandshakeResponse.class);
  private SpecificDatumReader<HandshakeRequest> handshakeReader =
      new SpecificDatumReader<HandshakeRequest>(HandshakeRequest.class);

  private Protocol handshake(Decoder in, Encoder out, Transceiver connection)
      throws IOException {
    if (connection != null && connection.isConnected())
      return connection.getRemote();
    HandshakeRequest request = (HandshakeRequest) handshakeReader.read(null, in);
    Protocol remote = protocols.get(request.clientHash);
    if (remote == null && request.clientProtocol != null) {
      remote = Protocol.parse(request.clientProtocol.toString());
      protocols.put(request.clientHash, remote);
    }
    HandshakeResponse response = new HandshakeResponse();
    if (localHash.equals(request.serverHash)) {
      response.match =
          remote == null ? HandshakeMatch.NONE : HandshakeMatch.BOTH;
    } else {
      response.match =
          remote == null ? HandshakeMatch.NONE : HandshakeMatch.CLIENT;
    }
    if (response.match != HandshakeMatch.BOTH) {
      response.serverProtocol = local.toString();
      response.serverHash = localHash;
    }

    RPCContext context = new RPCContext();
    context.setHandshakeRequest(request);
    context.setHandshakeResponse(response);
    for (RPCPlugin plugin : rpcMetaPlugins) {
      plugin.serverConnecting(context);
    }
    handshakeWriter.write(response, out);

    if (connection != null && response.match != HandshakeMatch.NONE)
      connection.setRemote(remote);

    return remote;
  }

  public CompletableFuture<Object> respondAsync(Message message, Object request) {
    CompletableFuture<Object> completableFuture = new CompletableFuture<Object>();
    try {
      completableFuture.complete(respond(message, request));
    } catch (Exception e) {
      completableFuture.completeExceptionally(e);
    }
    return completableFuture;
  }

  /** Reads a request message. */
  public abstract Object respond(Message message, Object request) throws Exception;

  /**
   * Reads a request message.
   */
  public abstract Object readRequest(Schema actual, Schema expected, Decoder in)
      throws IOException;

  /** Writes a response message. */
  public abstract void writeResponse(Schema schema, Object response,
                                     Encoder out) throws IOException;

  /** Writes an error message. */
  public abstract void writeError(Schema schema, Object error,
                                  Encoder out) throws IOException;

}

File: lang/php/lib/Ipc/Responder.php
<?php

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Avro\Ipc;

use Apache\Avro\AvroException;
use Apache\Avro\Datum\AvroIOBinaryDecoder;
use Apache\Avro\Datum\AvroIOBinaryEncoder;
use Apache\Avro\Datum\AvroIODatumReader;
use Apache\Avro\Datum\AvroIODatumWriter;
use Apache\Avro\IO\AvroStringIO;
use Apache\Avro\Protocol\AvroProtocol;
use Apache\Avro\Protocol\AvroProtocolMessage;
use Apache\Avro\Protocol\AvroProtocolParseException;
use Apache\Avro\Schema\AvroSchema;
use Apache\Avro\Schema\AvroSchemaParseException;

/**
 * Base class for the client side of a protocol interaction.
 * @package Avro
 */
abstract class Responder
{
    public const HANDSHAKE_REQUEST_SCHEMA_JSON = <<<HRSJ
    {
      "type": "record",
      "name": "HandshakeRequest", "namespace":"org.apache.avro.ipc",
      "fields": [
        {"name": "clientHash",
         "type": {"type": "fixed", "name": "MD5", "size": 16}},
        {"name": "clientProtocol", "type": ["null", "string"]},
        {"name": "serverHash", "type": "MD5"},
        {"name": "meta", "type": ["null", {"type": "map", "values": "bytes"}]}
      ]
    }
    HRSJ;

    public const HANDSHAKE_RESPONSE_SCHEMA_JSON = <<<HRSJ
    {
      "type": "record",
      "name": "HandshakeResponse", "namespace": "org.apache.avro.ipc",
      "fields": [
        {"name": "match",
         "type": {"type": "enum", "name": "HandshakeMatch",
                  "symbols": ["BOTH", "CLIENT", "NONE"]}},
        {"name": "serverProtocol",
         "type": ["null", "string"]},
        {"name": "serverHash",
         "type": ["null", {"type": "fixed", "name": "MD5", "size": 16}]},
        {"name": "meta",
         "type": ["null", {"type": "map", "values": "bytes"}]}
      ]
    }
    HRSJ;

    protected $local_protocol;
    protected $local_hash;
    protected $protocol_cache = array();

    protected $handshake_responder_writer;
    protected $handshake_responder_reader;
    protected $meta_writer;
    protected $meta_reader;

    protected $system_error_schema;

    public function __construct(AvroProtocol $local_protocol)
    {
        $this->local_protocol = $local_protocol;
        $this->local_hash = $local_protocol->md5();
        $this->protocol_cache[$this->local_hash] = $this->local_protocol;

        $this->handshake_responder_writer = new AvroIODatumWriter(
            AvroSchema::parse(Responder::HANDSHAKE_RESPONSE_SCHEMA_JSON)
        );
        $this->handshake_responder_reader = new AvroIODatumReader(
            AvroSchema::parse(Responder::HANDSHAKE_REQUEST_SCHEMA_JSON)
        );
        $this->meta_writer = new AvroIODatumWriter(AvroSchema::parse('{"type": "map", "values": "bytes"}'));
        $this->meta_reader = new AvroIODatumReader(AvroSchema::parse('{"type": "map", "values": "bytes"}'));

        $this->system_error_schema = AvroSchema::parse('["string"]');
    }

    public function local_protocol()
    {
        return $this->local_protocol;
    }

    /**
     * Entry point to process one procedure call.
     * @param string $call_request the serialized procedure call request
     * @param Transceiver $transceiver the transceiver used for the response
     * @return string|null the serialiazed procedure call response or null if it's a one-way message
     * @throws AvroException
     */
    public function respond($call_request, Transceiver $transceiver)
    {
        $buffer_reader = new AvroStringIO($call_request);
        $decoder = new AvroIOBinaryDecoder($buffer_reader);

        $buffer_writer = new AvroStringIO();
        $encoder = new AvroIOBinaryEncoder($buffer_writer);

        $error = null;
        $response_metadata = [];
        try {
            $remote_protocol = $this->process_handshake($decoder, $encoder, $transceiver);
            if (is_null($remote_protocol)) {
                return $buffer_writer->string();
            }

            $request_metadata = $this->meta_reader->read($decoder);
            $remote_message_name = $decoder->readString();
            if (!isset($remote_protocol->messages[$remote_message_name])) {
                throw new AvroException("Unknown remote message: $remote_message_name");
            }
            $remote_message = $remote_protocol->messages[$remote_message_name];

            if (!isset($this->local_protocol->messages[$remote_message_name])) {
                throw new AvroException("Unknown local message: $remote_message_name");
            }
            $local_message = $this->local_protocol->messages[$remote_message_name];

            $datum_reader = new AvroIODatumReader($remote_message->request, $local_message->request);
            $request = $datum_reader->read($decoder);
            try {
                $response_datum = $this->invoke($local_message, $request);
                // if it's a one way message we only send the handshake if needed
                if ($this->local_protocol->messages[$remote_message_name]->is_one_way()) {
                    return ($buffer_writer->string() === "") ? null : $buffer_writer->string();
                }

            } catch (AvroRemoteException $e) {
                $error = $e;
            } catch (\Exception $e) {
                $error = new AvroRemoteException($e->getMessage());
            }

            $this->meta_writer->write($response_metadata, $encoder);
            $encoder->writeBoolean(!is_null($error));

            if (is_null($error)) {
                $datum_writer = new AvroIODatumWriter($local_message->response);
                $datum_writer->write($response_datum, $encoder);
            } else {
                $datum_writer = new AvroIODatumWriter($remote_message->errors);
                $datum_writer->write($error->getDatum(), $encoder);
            }

        } catch (AvroException $e) {
            $error = new AvroRemoteException($e->getMessage());
            $buffer_writer = new AvroStringIO();
            $encoder = new AvroIOBinaryEncoder($buffer_writer);
            $this->meta_writer->write($response_metadata, $encoder);
            $encoder->writeBoolean(!is_null($error));
            $datum_writer = new AvroIODatumWriter($this->system_error_schema);
            $datum_writer->write($error->getMessage(), $encoder);
        }

        return $buffer_writer->string();
    }

    /**
     * Processes an RPC handshake.
     * @param AvroIOBinaryDecoder $decoder Where to read from
     * @param AvroIOBinaryEncoder $encoder Where to write to.
     * @param Transceiver $transceiver the transceiver used for the response
     * @return AvroProtocol The requested Protocol.
     * @throws AvroProtocolParseException|AvroSchemaParseException
     */
    public function process_handshake(
        AvroIOBinaryDecoder $decoder,
        AvroIOBinaryEncoder $encoder,
        Transceiver $transceiver
    ) {
        if ($transceiver->is_connected()) {
            return $transceiver->get_remote();
        }

        $handshake_request = $this->handshake_responder_reader->read($decoder);
        $client_hash = $handshake_request["clientHash"];
        $client_protocol = $handshake_request["clientProtocol"];
        $remote_protocol = $this->get_protocol_cache($client_hash);

        if (is_null($remote_protocol) && !is_null($client_protocol)) {
            $remote_protocol = AvroProtocol::parse($client_protocol);
            $this->set_protocol_cache($client_hash, $remote_protocol);
        }

        $server_hash = $handshake_request["serverHash"];
        $handshake_response = array();

        if ($this->local_hash == $server_hash) {
            $handshake_response['match'] = (is_null($remote_protocol)) ? 'NONE' : 'BOTH';
        } else {
            $handshake_response['match'] = (is_null($remote_protocol)) ? 'NONE' : 'CLIENT';
        }

        $handshake_response["meta"] = null;
        if ($handshake_response['match'] !== 'BOTH') {
            $handshake_response["serverProtocol"] = $this->local_protocol->__toString();
            $handshake_response["serverHash"] = $this->local_hash;
        } else {
            $handshake_response["serverProtocol"] = null;
            $handshake_response["serverHash"] = null;
        }

        $this->handshake_responder_writer->write($handshake_response, $encoder);

        if ($handshake_response['match'] !== 'NONE') {
            $transceiver->set_remote($remote_protocol);
        }

        return $remote_protocol;
    }

    /**
     * @param string $hash hash of an Avro Protocol
     * @return AvroProtocol|null The protocol associated with $hash or null
     */
    public function get_protocol_cache($hash)
    {
        return $this->protocol_cache[$hash] ?? null;
    }

    /**
     * @param string $hash hash of an Avro Protocol
     * @param AvroProtocol $protocol
     * @return Responder $this
     */
    public function set_protocol_cache($hash, AvroProtocol $protocol)
    {
        $this->protocol_cache[$hash] = $protocol;
        return $this;
    }

    /**
     * Processes one procedure call
     * @param AvroProtocolMessage $local_message
     * @param mixed $request Call request
     * @return mixed Call response
     */
    abstract public function invoke($local_message, $request);
}
